### PR TITLE
fix: improve `Var_t` expression equality check

### LINKED ISSUE #4131: bug: arguments do not match while selecting procedure from interface
```fortran
MODULE thrd_comm_module
  INTERFACE precv
    MODULE PROCEDURE precv_d_3d
  END INTERFACE precv
  CONTAINS

  SUBROUTINE precv_d_3d ( value )
    REAL, DIMENSION(4), INTENT(IN) :: value
  END SUBROUTINE precv_d_3d

  SUBROUTINE sweep_recv_bdry ( nz )

    INTEGER, INTENT(IN) ::  nz

    REAL, DIMENSION(nz):: jb_in

    CALL precv ( jb_in )

  END SUBROUTINE sweep_recv_bdry

END MODULE thrd_comm_module
```

```console
% lfortran -c thrd_comm.f90
semantic error: Arguments do not match for any generic procedure, precv
  --> thrd_comm.f90:19:5
   |
19 |     CALL precv ( jb_in )
   |     ^^^^^^^^^^^^^^^^^^^^ 


Note: Please report unclear, confusing or incorrect messages as bugs at
https://github.com/lfortran/lfortran/issues.
```

### PR DESCRIPTION
**Pull request by kmr-srbh**
fixes #4131 

### The issue
The chief cause of the issue was the check for comparing equality of `Var_t` expressions. We had been checking if two `Var_t` expressions were equal by checking if their symbols were equal. This comparison was not correct and always returned `false`, hence leading the compiler to believe that the argument passed to the subroutine did not match the parameter type. 

Here is the debug comment output for the MRE:

```fortran
MODULE thrd_comm_module
   INTERFACE precv
      MODULE PROCEDURE precv_d_3d
   END INTERFACE precv

CONTAINS

   SUBROUTINE precv_d_3d ( value, nx )

      REAL, DIMENSION(nx), INTENT(IN) :: value

      INTEGER, INTENT(IN) ::  nx

   END SUBROUTINE precv_d_3d

   SUBROUTINE sweep_recv_bdry ( nz )

      REAL, DIMENSION(nz):: jb_in

      INTEGER, INTENT(IN) ::  nz

      CALL precv ( jb_in, nz )

   END SUBROUTINE sweep_recv_bdry


END MODULE thrd_comm_module
```
```console
(lf) saurabh-kumar@Awadh:~/Projects/System/lfortran$ lfortran -c ./examples/example.f90
n_args: 2
procedure name: precv
p.n_procs: 1
else
is function
args.size(): 2
sub.n_args: 2
type1 real[:]
type2: real[:]
is array
n_dims_a: 1
n_dims_b: 1
x->type: integer
y->type: integer
var
var_x->m_v: nz
var_y->m_v: nx
expr not equal
types not equal
semantic error: Arguments do not match for any generic procedure, precv
  --> ./examples/example.f90:22:7
   |
22 |       CALL precv ( jb_in, nz )
   |       ^^^^^^^^^^^^^^^^^^^^^^^^ 


Note: Please report unclear, confusing or incorrect messages as bugs at
https://github.com/lfortran/lfortran/issues.
```

### The fix
We fix this by comparing `ASR::exprType` of the `Var_t` expressions to see if they are equal.

```console
(base) saurabh-kumar@Awadh:~/Projects/System/lfortran$ lfortran -c ./examples/example.f90
n_args: 2
procedure name: precv
p.n_procs: 1
else
is function
args.size(): 2
sub.n_args: 2
type1 real[:]
type2: real[:]
is array
n_dims_a: 1
n_dims_b: 1
x->type: integer
y->type: integer
var
var_x->m_v type: integer
var_y->m_v type: integer
dim_a_int: 1
dim_b_int: 1
type1 integer
type2: integer
```

With the fix in place we are now successfully able to compile the file `thrd_comm.f90` of the SNAP package.

```console
(base) saurabh-kumar@Awadh:~/Projects/Packages/SNAP/src$ lfortran -c thrd_comm.f90
(base) saurabh-kumar@Awadh:~/Projects/Packages/SNAP/src$
```

### CODE REVIEW

**Comment by Pranavchiku** [comment]
Let's get CI all passed.

**Comment by kmr-srbh** [comment]
@Pranavchiku the cause of failure here is a warning being treated as error on macOS with the AppleClang compiler I see that a proposal for supporting the above warning in GCC is open [here](https://gcc.gnu.org/bugzilla/show_bug.cgi?id=70181).

When the expression types are equal, it will essentially be a self-comparison, right? Could you suggest any ideas to bypass this, or do this in a better way?

File: src/libasr/asr_utils.h [hunk 1]
```diff
@@ -2998,7 +2998,7 @@ inline bool expr_equal(ASR::expr_t* x, ASR::expr_t* y) {
         case ASR::exprType::Var: {
             ASR::Var_t* var_x = ASR::down_cast<ASR::Var_t>(x);
             ASR::Var_t* var_y = ASR::down_cast<ASR::Var_t>(y);
-            return var_x->m_v == var_y->m_v;
+            return var_x->class_type == var_y->class_type;
```
Comment (Pranavchiku) [inline src/libasr/asr_utils.h:3001]:
```suggestion
            return ASRUtils::types_equal(var_x->class_type, var_y->class_type);
```

We do not compare types like this, refer `ASRUtils::types_equal` .

**Review by Pranavchiku — CHANGES_REQUESTED** [review]
left a comment above.

File: src/libasr/asr_utils.h (hunk 1 above)
Comment (kmr-srbh) [reply src/libasr/asr_utils.h]: Thanks for this @Pranavchiku! I could fix this using a similar function. Let's hope the CI passes this time.

**Comment by certik** [comment]
If this works, then go ahead and add an integration test for this.

**Comment by Pranavchiku** [comment]
Please add an integration test and wrap this PR as soon as possible.

File: src/libasr/asr_utils.h [hunk 2]
```diff
@@ -3432,7 +3434,7 @@ inline bool types_equal_with_substitution(ASR::ttype_t *a, ASR::ttype_t *b,
     return false;
 }
 
-inline bool check_equal_type(ASR::ttype_t* x, ASR::ttype_t* y, bool check_for_dimensions=false) {
+inline bool check_equal_type(ASR::ttype_t* x, ASR::ttype_t* y, bool check_for_dimensions) {
```
Comment (Pranavchiku) [inline src/libasr/asr_utils.h:3437]: Why do we need this change?

File: src/libasr/asr_utils.h [hunk 3]
```diff
@@ -2998,7 +3000,7 @@ inline bool expr_equal(ASR::expr_t* x, ASR::expr_t* y) {
         case ASR::exprType::Var: {
             ASR::Var_t* var_x = ASR::down_cast<ASR::Var_t>(x);
             ASR::Var_t* var_y = ASR::down_cast<ASR::Var_t>(y);
-            return var_x->m_v == var_y->m_v;
+            return check_equal_type(expr_type(&var_x->base), expr_type(&var_y->base));
```
Comment (Pranavchiku) [inline src/libasr/asr_utils.h:3003]: If you wish to test dimension as well do `check_equal_type(expr_type(&var_x->base), expr_type(&var_y->base), true)`

File: integration_tests/var_expression_comparison.f90 [hunk 4]
```diff
@@ -0,0 +1,85 @@
...
+!    END SUBROUTINE subroutine_2   
+!
+!
+! the expression type of the argument variable `ny` to `subroutine_1` must be equal to 
+! the parameter variable `nx`.
+
+
+MODULE test_var_expression_comparison_module
+
+CONTAINS
```
Comment (Pranavchiku) [inline integration_tests/var_expression_comparison.f90:42]: Also, I think if we do not have interface then this is different issue and not what we need for SNAP.

**Review by Pranavchiku — COMMENTED** [review]
left questions above

File: src/libasr/asr_utils.h (hunk 2 above)
Comment (kmr-srbh) [reply src/libasr/asr_utils.h:3437]: We were unable to use this function as it is defined later in the code. To use it here, I have declared it at the top.

File: src/libasr/asr_utils.h (hunk 3 above)
Comment (kmr-srbh) [reply src/libasr/asr_utils.h:3003]: Okay.

File: src/libasr/asr_utils.h (hunk 2 above)
Comment (Pranavchiku) [reply src/libasr/asr_utils.h:3437]: I mean to ask why to remove `bool check_for_dimensions=false`

File: integration_tests/var_expression_comparison.f90 (hunk 4 above)
Comment (kmr-srbh) [reply integration_tests/var_expression_comparison.f90:42]: Ah, right! I am making the changes,

File: src/libasr/asr_utils.h (hunk 2 above)
Comment (kmr-srbh) [reply src/libasr/asr_utils.h:3437]: We cannot have them simultaneously in the declaration and definition. The declaration contains the default value.

Comment (Pranavchiku) [reply src/libasr/asr_utils.h:3437]: Why is it working in latest main then?

File: integration_tests/CMakeLists.txt [hunk 5]
```diff
@@ -1574,3 +1574,4 @@ RUN(NAME module_function_with_nopass LABELS gfortran llvm llvm_wasm llvm_wasm_em
 RUN(NAME module_function_without_nopass LABELS gfortran llvm llvm_wasm llvm_wasm_emcc)
 
 RUN(NAME polymorphic_argument LABELS gfortran llvm llvm_wasm llvm_wasm_emcc)
+RUN(NAME var_expression_comparison LABELS gfortran llvm llvm_wasm llvm_wasm_emcc)
```
Comment (Pranavchiku) [inline integration_tests/CMakeLists.txt:1577]:
```suggestion
RUN(NAME interface_01 LABELS gfortran llvm llvm_wasm llvm_wasm_emcc)
```

I'll prefer name like this.

File: src/libasr/asr_utils.h (hunk 2 above)
Comment (kmr-srbh) [reply src/libasr/asr_utils.h:3437]: It is so because before this, we had not used the function in the same file `asr_utils.h` before it's definition.

Comment (Pranavchiku) [reply src/libasr/asr_utils.h:3437]: I see, you added signature above, good.

**Review by Pranavchiku — COMMENTED** [review]
changes are good, would you like to clean commit history or squash and merge would be fine?

**Comment by kmr-srbh** [comment]
> changes are good, would you like to clean commit history or squash and merge would be fine?

Let's squash and merge this as a single commit.

**Comment by kmr-srbh** [comment]
Thank you very much for the other test @Pranavchiku! Thank you very much for guiding and supporting me!

**Comment by Pranavchiku** [comment]
Go ahead, pull changes, scan SNAP, create MRE and march to fix it.

**Pull request merged**